[XLA] Simplify transposes that are really reshapes

Keno opened the pull request:
A transpose like
```
f32[1,1,64,1] = transpose(f32[1,64,1,1]), dimensions={3,2,1,0}
```
is really just a reshape (because there's only one non-1 dimension).
Teach algebraic simplifier to make that substitution, to enable applying
reshape-combining optimizations to such instructions.

Keno committed:
[XLA] Simplify transposes that are really reshapes

A transpose like
```
f32[1,1,64,1] = trasnpose(f32[1,64,1,1]), dimensions={3,2,1,0}
```
is really just a reshape (because there's only one non-1 dimension).
Teach algebraic simplifier to make that substitution, to enable applying
reshape-combining optimizations to such instructions.

Keno commented:
This change gives about a 15% performance improvement on TPUs when running ResNet50 as implemented in https://github.com/FluxML/MetalHead.jl.

Keno commented:
Addressed review comment from @davidel.

tensorflowbutler commented:
Nagging Assignee @aaroey: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

blakehechtman commented on tensorflow/compiler/xla/service/algebraic_simplifier.cc:
```diff
@@ -1875,10 +1875,13 @@ Status AlgebraicSimplifierVisitor::HandleReshape(HloInstruction* reshape) {
-    return ReplaceWithNewInstruction(
-        reshape, HloInstruction::CreateReshape(reshape->shape(),
-                                               operand->mutable_operand(0)));
+    HloInstruction* new_reshape =
```
why this change?

blakehechtman commented on tensorflow/compiler/xla/service/algebraic_simplifier.cc:
```diff
@@ -2520,6 +2523,13 @@ Status AlgebraicSimplifierVisitor::HandleTranspose(HloInstruction* transpose) {
+    HloInstruction* reshape = computation_->AddInstruction(
+        HloInstruction::CreateReshape(transpose->shape(), operand));
+    TF_RETURN_IF_ERROR(ReplaceInstruction(transpose, reshape));
+    return HandleReshape(reshape);
```
algebraic simplifier runs in a fixed point pass, there is not a reason to recurse here?

blakehechtman commented on tensorflow/compiler/xla/service/algebraic_simplifier_test.cc:
```diff
@@ -1770,6 +1770,28 @@ TEST_F(AlgebraicSimplifierTest, TransposesMerged) {
+
+  AlgebraicSimplifier simplifier(/*is_layout_sensitive=*/false,
+                                 bitcasting_callback());
+  EXPECT_TRUE(simplifier.Run(module.get()).ValueOrDie());
```
If you get rid of the recursion yo and execute this many times.

blakehechtman commented on tensorflow/compiler/xla/service/algebraic_simplifier.cc:
```diff
@@ -2520,6 +2523,13 @@ Status AlgebraicSimplifierVisitor::HandleTranspose(HloInstruction* transpose) {
                                            transpose->dimensions())));
   }
 
+  if (ShapeUtil::TrueRank(transpose->shape()) <= 1) {
```
Can you generalize this to Detect degenerate dimension, strip them from the transpose and then make remove ones reshape, transpose, add ones reshape. At that point the tranpose becomes a no-op and the reshapes are free to merge in the next fixed point through the algebraic simplifier.

Keno commented:
Updated to remove the recursion and rely on fix pointing instead. I'll take a look at the more general transform.

Keno committed:
[XLA] Canonicalize Transpose by dropping degenerate dims

This generates extra reshapes, but hopefully they should be able
to be folded into the surrounding instructions or otherwise be
free.

Keno commented:
Added the more general thing as a separate commit.

blakehechtman commented on tensorflow/compiler/xla/service/algebraic_simplifier.cc:
```diff
@@ -2748,6 +2749,27 @@ Status AlgebraicSimplifierVisitor::HandleSort(HloInstruction* sort) {
+  int64 i = 0;
+  for (int64 sz : shape.dimensions()) {
+    if (sz == 1) {
+      dimension_map.push_back(-1);
```
spell out dim_size

blakehechtman commented on tensorflow/compiler/xla/service/algebraic_simplifier_test.cc:
```diff
@@ -2047,6 +2047,27 @@ TEST_F(AlgebraicSimplifierTest, TransposesMerged) {
+    HloModule module
+
+    ENTRY test {
+      param = f32[10] parameter(0)
```
add a test with more than one non-trivial dimension

Keno commented:
Before this gets merged, I want to add the test that Blake requested and fix the typo. Will have that shortly, just one sec.

Keno commented:
Added (and fixed a small bug that Blake pointed out to me in person yesterday).

Keno commented:
Thanks!
